packages/gatsby-worker/src/task-queue.ts

```diff
@@ -0,0 +1,49 @@
+interface ITaskQueueNode<ValueType> {
+  value: ValueType
+  next?: ITaskQueueNode<ValueType>
+  prev?: ITaskQueueNode<ValueType>
+}
+export class TaskQueue<ValueType> {
+  private head?: ITaskQueueNode<ValueType>
+  private tail?: ITaskQueueNode<ValueType>;
+
+  *[Symbol.iterator](): Iterator<ITaskQueueNode<ValueType>> {
+    let currentHead = this.head
+    while (currentHead) {
+      yield currentHead
+      currentHead = currentHead.next
+    }
+  }
+
+  enqueue(task: ValueType): void {
+    const newNode: ITaskQueueNode<ValueType> = {
+      value: task,
+    }
+    if (this.tail) {
+      this.tail.next = newNode
+      newNode.prev = this.tail
+    } else {
+      this.head = newNode
+    }
+
+    this.tail = newNode
+  }
+
+  remove(taskNode: ITaskQueueNode<ValueType>): void {
+    if (taskNode === this.head) {
+      this.head = taskNode.next
+      if (this.head) {
+        this.head.prev = undefined
+      } else {
+        // if we don't have head, we also don't have tail
+        this.tail = undefined
+      }
+    } else {
+      if (taskNode === this.tail) {
+        this.tail = taskNode.prev
+      }
+      // if node is not the head then it will have .prev
+      taskNode.prev!.next = taskNode.next
```

Comment:
Should we also re-assign `taskNode.next.prev`? Is it possible that it will retain the task through this link?

Reply:
Yes, you are correct - damn basic data structures ;)

Reply:
done in https://github.com/gatsbyjs/gatsby/pull/32120/commits/14741b9ed9b3b6af875d3372b2659315bebda1ec